File: CLEAR_Matlab/CLEAR.m
%% CLEAR algorithm (Consistent Laplacian Estimatmion and Aberrancy Reduction)
%
% Inputs:
%
%           - TT:       Initial permutations (aka. correspondences or score matrices)
%           - numSmp:   Number of observations for each agent
%           - numAgt:   Number of agents (aka. frames or observations)
%
% Outputs:
%
%           - XX:       Consistent pariwise permutations
%           - X:        Map to universe (lifting permutations)
%           - numObj:   Estimated number of objects
%
% Options:
%
%           - 'numObj':     When the number of objects in the universe is
%                           provided. E.g., (..., 'numObj', 10)
%                           If 'numObj' is not specified it will be estimated
%                           from the sepctrum of Laplacian automatically. s
%
%
% If this program is useful, please cite:
%
% [1] K. Fathian, K. Khosoussi, P. Lusk, Y. Tian, J.P. How, "CLEAR: A 
%     Consistent Lifting, Embedding, and Alignment Rectification Algorithm 
%     for Multi-Agent Data Association", arXiv:1902.02256, 2019.
%
%
% This program is free software: you can redistribute it and/or modify it
% under the terms of the GNU lesser General Public License, either version 
% 3, or any later version. This program is distributed in the hope that it 
% will be useful, but WITHOUT ANY WARRANTY. 
%
% (C) Kaveh Fathian, 2019.  Email: kaveh.fathian@gmail.com
%
%
%% Changes:
%
%
function [XX, X, numObj] = CLEAR(TT, numSmp, numAgt, varargin)
%% Parse input

% Preallocate paramters
numObj = -1; 

ivarargin = 1;
while ivarargin <= length(varargin)
    switch lower(varargin{ivarargin})
        case 'numobj'
            ivarargin = ivarargin+1;
            numObj = varargin{ivarargin};        
        otherwise
            fprintf('Unknown option ''%s'' is ignored!',varargin{ivarargin});
    end
    ivarargin = ivarargin+1;
end

idxSum = [0, cumsum(numSmp)]; % Cumulative index
numSmpSum = sum(numSmp); % Total number of observations

if ~(numSmpSum == size(TT,1))
    error('Incorrect number of samples.')
end


%% %%%%%%%%%%%%%%%%%%%%%%%%%%%% CLEAR algorithm

P = (TT + TT')/2; % Make association matrix symmetric
for i = 1 : numAgt % Remove any self associations (distinctness constraint)
    idxi = [idxSum(i)+1 : idxSum(i+1)];
    P(idxi,idxi) = eye(numSmp(i)); % Block of P associated to agent i
end
A = P - diag(diag(P)); % Adjacency matrix of induced graph
L = P2L(P); % Graph Laplacian matrix


% Normalize L
Lnrm = NormalizeLap(L, 'normalize', 'DI', 'multtype', 'sym');


% Compute SVD using union of connected components' SVDs (to improve speed)
[sl,Vl] = BlockSVD(A, Lnrm);
% [~,Sl,Vl] = svd(Lnrm); sl = diag(Sl);


% Estimate size of universe if not provided
if numObj == -1
    numObj = EstimateNumObj(sl, 'eigval',true, 'numsmp',numSmp);
end


% Get the null space
U0 = Vl(:, end-numObj+1 : end); % Kernel
U = bsxfun(@rdivide, U0, sqrt(sum(U0.^2,2))); % Normalize each row of U0


% Find cluster center candidates (Can be improved??)
C = PivotRows(U, numObj); % Cluster centers


% Distance to cluster centers
F = 1 - U * C'; % Dot product with cluster centers (the minimum element indicates the optimal assignment)    


% Solve linear assignment
X = zeros(size(F));
for i = 1 : numAgt
    idxi = [idxSum(i)+1 : idxSum(i+1)];
    Fi = F(idxi, :); % Component of F associated to agent i
    
    % suboptimal linear assignment
    Xi = SuboptimalAssign(Fi);

    X(idxi, :) = Xi; % Store results
end


% Pairwise assignments
XX = X * X';

File: CLEAR_Python/CLEAR.py
import numpy as np
import util

def CLEAR(TT, numSmp, numAgt, numObj=None):
    """
    CLEAR algorithm (Consistent Laplacian Estimatmion and Aberrancy Reduction)

    Parameters
    ----------
    TT : numpy.ndarray
        Initial permutations (aka. correspondences or score matrices)
    numSmp : numpy.ndarray
        Number of observations for each agent
    numAgt : int
        Number of agents (aka. frames or observations)
    numObj : int, optional
        Number of objects in the universe. The default is None. If None, 
        numObj will be estimated from the sepctrum of Laplacian automatically.
    
    Returns
    -------
    XX : numpy.ndarray
        Consistent pairwise permutations
    X : numpy.ndarray
        Map to universe (lifting permutations)
    numObj : int
        Estimated number of objects
    """
    
    # Also work with regular array input
    TT = np.array(TT)
    numSmp = np.array(numSmp)
    
    idxSum = np.append([0], np.cumsum(numSmp)) # Cumulative index
    numSmpSum = np.sum(numSmp) # Total number of observations
    
    if numSmpSum != TT.shape[0]:
        raise ValueError("Incorrect number of samples.")
    
    P = (TT + TT.T)/2 # Make association matrix symmetric
    for i in range(numAgt): # Remove any self associations (distinctness constraint)
        l,r = idxSum[i], idxSum[i+1]
        P[l:r, l:r] = np.eye(numSmp[i]) # Block of P associated to agent i
        
    A = P - np.diag(np.diag(P)) # Adjacency matrix of induced graph
    L = util.P2L(P) # Graph Laplacian matrix
    
    # Normalize L
    Lnrm = util.normalize_lap(L, "DI", "sym")
    
    # Compute SVD using union of connected components' SVDs (to improve speed)
    sl, Vl = util.block_svd(A, Lnrm)

    # Estimate size of universe if not provided
    if numObj == None:
        numObj, _ = util.estimate_num_obj(sl, eigval=True, numSmp=numSmp)
        
    # Get the null space
    U0 = Vl[:, -numObj:] # Kernel
    U = U0/np.linalg.norm(U0, axis=1, keepdims=True) # Normalize each row of U0
    
    # Find cluster center candidates
    C = util.pivot_rows(U, numObj)
    
    # Distance to cluster centers
    F = 1 - np.matmul(U, C.T)

    # Solve linear assignment
    X = np.zeros(F.shape)
    for i in range(numAgt):
        l,r = idxSum[i], idxSum[i+1]
        Fi = F[l:r, :] # Component of F associated to agent i
        
        # Suboptimal linear assignment
        Xi = util.suboptimal_assignment(Fi)
        
        X[l:r, :] = Xi # Store results
        
    # Pairwise assignments
    XX = np.matmul(X, X.T)
    
    return XX, X, numObj